framework/src/tinyalsa/tinyalsa.c

```diff
@@ -561,7 +562,7 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 		return -EINVAL;
 	}
 
-	if (!(pcm->flags & PCM_IN)) {
+	if (!(pcm->flags & PCM_IN) || pcm->flags & PCM_MMAP) {
```

Comment:
i think, pcm->flags & PCM_OUT is better.

Reply:
i mean,
if ((pcm->flags & PCM_OUT) || pcm->flags & PCM_MMAP)

Reply:
Since it is pcm_read API, I think its better to check for PCM_IN. 
The if-statement is a bit confusing, but it works well.. and maybe more logically correct.

Reply:
I think there's no change call pcm_readi with PCM_MMAP flags, you can remove second condition

Reply:
Ok. Done.